Profile page performance issue Fix

saadmk11 opened the pull request:
closes #5468

safwanrahman commented on readthedocs/core/templatetags/privacy_tags.py:
```diff
@@ -18,9 +20,13 @@ def is_admin(user, project):
 
 @register.simple_tag(takes_context=True)
 def get_public_projects(context, user):
+    builds = Build.objects.filter(project=OuterRef('pk'))
```
Can you add some comments here describing the `subquery` and others and why have you gone this approach?

saadmk11 replied:
okay

safwanrahman commented:
This seems like a elegant fix for the optimization. I did not know about solving the perfomance issue using the `Exits` query.

saadmk11 commented:
@davidfischer I would like to give it a try if its okay with you.

saadmk11 commented:
@davidfischer Updated The PR please have a look and let me know if the implementation is what you wanted. and what do you think I should do for the other annotation `latest_build_date` as it is used in `/projects/tags/tag-slug/` page also.

davidfischer commented:
I like the solution for `has_good_build` and that was exactly what I was thinking!

I think we can use a similar approach for `latest_build_date`. Perhaps there's a way to get the whole build object and re-use the property `project.get_latest_build`?

saadmk11 commented:
@davidfischer Great! Actually I Didn't use `get_latest_build` because it created an extra Database query. But now following this approach  we can solve it. I'll update the PR soon.

davidfischer commented:
This was a little trickier than I thought but you can control the prefetching with a `Prefetch` object:

```python
from readthedocs.projects.models import Project
from readthedocs.builds.models import Build

from django.db.models import Prefetch


latest_build = Prefetch('builds', queryset=Build.objects.filter(state='finished', type='html'), to_attr='_latest_build')

# Executes 2 queries
projects = list(Project.objects.all().prefetch_related(latest_build))

# This executes 0 queries
for project in projects:
    print(project.slug)
    if project._latest_build:
        # It is a little awkward that "_latest_build" is a list... and it could be empty
        print(' - {}'.format(project._latest_build[0].date))
```

The above code isn't ideal but it should give you some tips on working this out.

safwanrahman commented:
@davidfischer The queryset `projects = list(Project.objects.all().prefetch_related(latest_build))` will fetch all the builds from all the projects that are selected in second query. I think its more likely memory heavy and make the queryset also much slow.

Do you think its good enough for fetching the latest build data that we need or fetch the whole object?

davidfischer commented:
> I think its more likely memory heavy and make the queryset also much slow.

Hmmmm, you're right. Maybe there's a way to limit it to just 1 result.

davidfischer commented:
> Do you think its good enough for fetching the latest build data that we need or fetch the whole object?

A lot of things get simpler if we can fetch the whole build object. However, we don't need to fetch 100s of build objects if possible.

davidfischer commented:
Thanks for looking at this @safwanrahman. I think this will do it if you wouldn't mind taking another look:

```python
from django.db.models import OuterRef, Subquery, Prefetch

from readthedocs.projects.models import Project
from readthedocs.builds.models import Build


subquery = Subquery(Build.objects.filter(state='finished', type='html', project=OuterRef('project_id')).values_list('id', flat=True)[:1])
latest_build = Prefetch('builds', Build.objects.filter(pk__in=subquery), to_attr='_latest_build')
projects = list(Project.objects.all().prefetch_related(latest_build))
```

saadmk11 commented:
@davidfischer Thanks a lot for your help. I tried out your code in the shell and also in the application but I can't seem to reduce the number of Database Queries. When ever `project._latest_build` is called it creates a new Query to the database.

```python
>>> from django.db import connection
>>> from django.db.models import OuterRef, Subquery, Prefetch
>>> from readthedocs.projects.models import Project
>>> from readthedocs.builds.models import Build
>>> 
>>> subquery = Subquery(Build.objects.filter(state='finished', type='html', project=OuterRef('project_id')).values_list('id', flat=True)[:1])
>>> latest_build = Prefetch('builds', Build.objects.filter(pk__in=subquery), to_attr='_latest_build')
>>> # Executes 2 queries
>>> projects = list(Project.objects.all().prefetch_related(latest_build))
>>> 
>>> # Executes 8 queries as 8 projects had builds.
>>> for project in projects:
...     if project._latest_build:
...         project._latest_build
... 
[<Build: Build Cue for AnonymousUser admin (11)>]
[<Build: Build Django Test Utils for AnonymousUser admin (13)>]
[<Build: Build Fabric for AnonymousUser admin (15)>]
[<Build: Build Kong for AnonymousUser admin (21)>]
[<Build: Build Pinax for AnonymousUser admin (19)>]
[<Build: Build Read The Docs for AnonymousUser admin (9)>]
[<Build: Build South for AnonymousUser admin (20)>]
[<Build: Build Tastypie for AnonymousUser admin (16)>]
>>> len(connection.queries)
10
```

davidfischer commented:
Those queries are coming from querying the `project_user` table from the [`__str__` method on build](https://github.com/rtfd/readthedocs.org/blob/44951fbf8fe1fe12fa9f99f3075f87bd77188525/readthedocs/builds/models.py#L619-L628). (the "for AnonymousUser" part). If you print say only the build date and ID you won't see them.

saadmk11 commented:
@davidfischer I have Updated the PR according to your recommendations.
I have removed `state='finished', type='html'` from the filter as previously the template wasn't using the `project.get_latest_build()` method it was just getting the latest build like `project.builds.all.0`. Let me know what you think about the PR and thanks a lot for Guiding me. :+1:

saadmk11 commented:
Great

davidfischer commented:
I'm interested in getting this merged soon if possible.
